py/server/deephaven/dtypes.py

```diff
@@ -286,3 +315,60 @@ def from_np_dtype(np_dtype: Union[np.dtype, pd.api.extensions.ExtensionDtype]) -
             return dtype
 
     return PyObject
+
+
+_numpy_int_type_codes = ["i", "l", "h", "b"]
+_numpy_floating_type_codes = ["f", "d"]
+
+
+def _from_numpy_scalar_to_py(x):
+    """Convert a numpy scalar to a Python scalar."""
+    if hasattr(x, "dtype"):
+        if x.dtype.char in _numpy_int_type_codes:
+            return int(x)
+        elif x.dtype.char in _numpy_floating_type_codes:
+            return float(x)
+        elif x.dtype.char == '?':
+            return bool(x)
+        elif x.dtype.char == 'U':
+            return str(x)
+        elif x.dtype.char == 'O':
+            return x
+        elif x.dtype.char == 'M':
+            from deephaven.time import to_j_instant
```

Comment:
Any reason this can't be at the top level?  If so, does it make things slow or is this optimized in python?

Reply:
That would lead to circular imports.